boards/DIM/Src/App/App_RegenPaddle.c

```diff
@@ -0,0 +1,74 @@
+#include <stdlib.h>
+#include <assert.h>
+#include "App_RegenPaddle.h"
+
+struct RegenPaddle
+{
+    uint32_t (*get_paddle_position)(void);
+    uint32_t lower_deadzone;
+    uint32_t upper_deadzone;
+    uint32_t paddle_position;
+    uint32_t regen;
+};
+
+struct RegenPaddle *App_RegenPaddle_Create(
+    uint32_t (*const get_pedal_position)(void),
+    uint32_t lower_deadzone,
+    uint32_t upper_deadzone)
+{
+    struct RegenPaddle *regen_paddle = malloc(sizeof(struct RegenPaddle));
+
+    assert(regen_paddle != NULL);
+
+    regen_paddle->get_paddle_position = get_pedal_position;
+    regen_paddle->lower_deadzone      = lower_deadzone;
+    regen_paddle->upper_deadzone      = upper_deadzone;
+
+    return regen_paddle;
+}
+
+void App_RegenPaddle_Destroy(struct RegenPaddle *const regen_paddle)
+{
+    free(regen_paddle);
+}
+
+ErrorCode App_RegenPaddle_Tick(struct RegenPaddle *const regen_paddle)
+{
+    const uint32_t paddle_position = regen_paddle->get_paddle_position();
+
+    if (paddle_position > 100)
+    {
+        return ERROR_CODE_OUT_OF_RANGE;
+    }
+
+    regen_paddle->paddle_position = paddle_position;
+
+    if (regen_paddle->paddle_position > regen_paddle->upper_deadzone)
+    {
+        regen_paddle->regen = 100;
+    }
+    else if (regen_paddle->paddle_position < regen_paddle->lower_deadzone)
+    {
+        regen_paddle->regen = 0;
+    }
+    else
+    {
+        regen_paddle->regen =
+            100 *
+            (regen_paddle->paddle_position - regen_paddle->lower_deadzone) /
+            (regen_paddle->upper_deadzone - regen_paddle->lower_deadzone);
+    }
+
+    return ERROR_CODE_OK;
+}
```

Comment:
Why is this a `tick` function? Why not `ErrorCode App_RegenPaddle_GetPaddlePosition(struct RegenPaddle* paddle, unint32_t* returned_paddle_position)`?

Reply:
Hmm maybe I am misunderstanding what `Tick` is but I think of it as a step to update the internal states (e.g. `paddle_position` and `regen`) using the underlying IO functions (e.g. `get_pedal_position`).

Whereas `GetPosition` and `GetRegen` just returns the value the internal state, which is essentially the calculations from the most recent update step. Does that line of reasoning seem sane to you?

Reply:
I mean, I understand it, but IMO we should just have `GetPosition` and `GetRegen` call the io functions directly and not have a `tick` function, because the abstraction around the padel doesn't actually have anything to do with ticking, that's a hardware-specific detail.

Reply:
To clarify, this would mean removing these internal variables right? Because we can just call the underlying io function each time to get the value.
```
uint32_t paddle_position;
uint32_t regen;
```
Also what kind of application would require a `tick` function then? Just for my future reference.

Reply:
>To clarify, this would mean removing these internal variables right? Because we can just call the underlying io function each time to get the value.

Yep.

> Also what kind of application would require a tick function then? Just for my future reference.

Hmm, I suppose anything that needs to run periodically. IMO this isn't periodic because we ask it when we want information. In general I avoid tick functions unless needed because they explicitly rely on side effects (generally) which makes them much harder to reason about and test.

Reply:
I guess getter functions that request an underlying hardware driver (e.g. ADC for regen paddle) for information should probably never have a tick function? I may have told @pasitlaothamatas to use a Tick for flow meter (which reads PWM frequency). I guess we should remove that `Tick` too? 

https://github.com/UBCFormulaElectric/Consolidated-Firmware/blob/master/boards/FSM/Src/App/App_FlowMeter.c

Reply:
Yeah, in general it makes things easier to misuse, because now all the users of whatever code you built need to understand that you must first call `tick` before calling any of the getter functions in order to get new information. It's not much, but it is mental overhead that makes it easier to create bugs.

Reply:
Awesome, I'll fix up flow meter later then :+1: